app/models/course/experience_points_record.rb

```diff
@@ -30,4 +32,34 @@ def manually_awarded?
   def reason
     manually_awarded? ? super : specific.experience_points_display_reason
   end
+
+  private
+
+  def send_notification
+    Course::LevelNotifier.level_reached(course_user.user, level_after_update)
+  end
+
+  # Test if the course_user will reach a new level after current update.
+  def reached_new_level?
+    return false if points_awarded.nil? || !points_awarded_changed?
+
+    current_level = level_before_update
+    new_level = level_after_update
+
+    current_level.persisted? &&
+      new_level.persisted? &&
```

Comment:
What we can do is to get the current level, then check next level's threshold here? Then we can do away with `level_before_update`. 

```ruby
def reached_new_level?
  return false if points_awarded.nil? || !points_awarded_changed?
  experience_points = course_user.experience_points
  current_level = course.level_for(experience_points)
  (experience_points + points_awarded) > current_level.next_level_threshold
end
```